enhancements/storage/storage-performant-security-policy.md

```diff
@@ -0,0 +1,202 @@
+
+This will speed up time spent while recursively changing selinux policies of a volume.
+
+By allowing cluster-admins to set namespace default of `selinuxChangePolicy` we want to give control over how these changes are applied to Openshift clusters, without explicitly  changing pods themselves.
+
+### User Stories
+
+N/A
+
+### Goals
+
+- Allow Openshift admins to configure per-namespace default of `fsGroupChangePolicy`.
+- Allow Openshift admins to configure per-namespace default of `selinuxChangePolicy`.
+
+### Non-Goals
+
+* Change Openshift default globally.
+
+## Proposal
+
+### Allow admins to opt-in to `fsGroupChangePolicy` via namespace label
+
+We propose a label `storage.openshift.io/fsgroup-change-policy` if that is set to `OnRootMismatch` in a namespace, then Openshift's SCC hooks start defaulting all pods created in that namespace with `OnRootMismatch` `FSGroupChangePolicy`.
+
+```go
+func getPodFsGroupChangePolicy(ns *corev1.Namespace) *api.PodFSGroupChangePolicy {
+    fsGroupPolicy, ok := ns.Labels[securityv1.FSGroupChangePolicy]
+    if !ok {
+        return nil
+    }
+    if fsGroupPolicy == "OnRootMismatch" {
+        onRootMismatchPolicy := api.FSGroupChangeOnRootMismatch
+        return &onRootMismatchPolicy
+    }
+    return nil
+}
+```
+
+### Allow admins to opt-in to `selinuxChangePolicy` via namespace label
+
+Similar to mechanism proposed above,  we propose to use label `storage.openshift.io/selinux-change-policy` to define namespace wide `selinuxChangePolicy` if none are specified in a pod.
+
+### Setting of `fsGroupChangePolicy` and `selinuxChangePolicy` during admission
+
+We propose to add a new admission hook (which will be eventually replaced with MAP) which will trigger the defautling logic, similar to how SCC matching hooks are triggered.
+
+### Eventual replacement with Mutating Admission policy
+
+We want to replace admission plugin with Mutating Admission Policy once the feature becomes generally available. Once Mutating Admission Policy becomes GA, we propose to replace the admission hook with MAP on the lines of https://docs.google.com/document/d/13xa7imyvQzok3IJ66mheGqemQGhkNySl-He1h1iRuSM/edit?usp=sharing
+
+This would serve similar purpose as admission hook we are carrying as go code.
+
+### Proof of concept
+
+We have already implemented a proof-of-concept https://github.com/openshift/kubernetes/pull/2311/files, which works and does the job. But this uses existing SCC admission hook.
+
+
+### Workflow Description
+
+Admins that want to set defaults of `fsGroupChangePolicy` and `selinuxChangePolicy` will set `storage.openshift.io/fsgroup-change-policy` or `storage.openshift.io/selinux-change-policy`
+labels in corresponding namespaces.
+
+This will result in setting these fields in all pods in affected namespace if these fields are *not* already set.
```

Comment:
Is there a way to choose a cluster wide default? That would be overridden by the namespace level label?

Reply:
No, I do not think there is a way to specify cluster wide default. That is explicitly mentioned as a non-goal.